src/v/serde/json/detail/numeric.cc

```diff
@@ -0,0 +1,275 @@
+                pos += 1;
+                continue;
+            } else if (buf[pos] == 'e' || buf[pos] == 'E') {
+                pos += 1;
+                _state = state::exponent_sign;
+                continue;
+            } else {
+                err = result::done;
+                _state = state::finished_with_double;
+
+                _double_acc = strtod_normal_precision(
+                  _double_acc, _exp_frac + _exp * (_exp_negative ? -1 : 1));
+                if (_double_acc > std::numeric_limits<double>::max()) {
+                    err = result::invalid_json_string;
+                    _state = state::finished_with_error;
+                    return pos;
+                }
+                if (_is_negative) {
+                    _double_acc = -_double_acc;
+                }
+
+                return pos;
+            }
+
+        case state::exponent_sign:
+            if (!_as_double) {
+                _as_double = true;
+                _double_acc = static_cast<double>(_u64_acc);
+            }
+
+            if (buf[pos] == '-') {
+                pos += 1;
+                _exp_negative = true;
+                _max_exp = (_exp_frac + 2147483639) / 10;
+            } else {
+                if (buf[pos] == '+') {
+                    pos += 1;
+                }
+
+                _max_exp = 308 - _exp_frac;
+            }
+            _state = state::exponent_first_digit;
+            continue;
+
+        case state::exponent_first_digit:
+            if (buf[pos] >= '0' && buf[pos] <= '9') {
+                _state = state::exponent_digit;
+                [[fallthrough]];
+            } else {
+                err = result::invalid_json_string;
+                _state = state::finished_with_error;
+                return pos;
+            }
+
+        case state::exponent_digit:
+            if (buf[pos] >= '0' && buf[pos] <= '9') {
+                bool skip = true;
+                if (_exp <= _max_exp) {
+                    _exp = _exp * 10 + static_cast<unsigned>(buf[pos] - '0');
+                    pos += 1;
+                    skip = false;
+                }
+
+                if (_exp > _max_exp) {
+                    if (_exp_negative) {
+                        // we are close to 0, drop the rest of the digits
+                        pos += skip;
+                        continue;
+                    } else {
+                        // we are maxing out the double to a number we won't be
+                        // able to represent in a double. We should stop
+                        // parsing.
+                        err = result::invalid_json_string;
+                        _state = state::finished_with_error;
+                        return pos;
+                    }
+                }
+                continue;
+            } else {
+                err = result::done;
+                _state = state::finished_with_double;
+
+                _double_acc = strtod_normal_precision(
```

Comment:
Another option to consider (I am not sure if this leads to better or worse precision / performance): parse as an absl::int128 and keep trace of the decimal position as the scale then convert to a float. There is plenty of precision available in a decimal 128 value that it should be good. I had to do something similar here: https://github.com/redpanda-data/connect/blob/0181223dd18e755fabaa25890f5f75a71ad542a1/internal/impl/snowflake/streaming/int128/decimal.go#L284